<h2> Gene Finder Final </h2>

<blockquote>
junhyun-nam/gene_finder.py
</blockquote>
<h3> Great Job! I have a few comments about some parts of your code below. </h3>
<pre><code class="language-python">


# -*- coding: utf-8 -*-
"""
Created on Sun Feb  2 11:24:42 2014

@author: Junhyun Nam

"""

# you may find it useful to import these variables (although you are not required to use them)
from amino_acids import aa, codons, aa_table
import random
from load import load_seq

def shuffle_string(s):
    """ Shuffles the characters in the input string
        NOTE: this is a helper function, you do not have to modify this in any way """
    return ''.join(random.sample(s,len(s)))

### YOU WILL START YOUR IMPLEMENTATION FROM HERE DOWN ###


def get_complement(nucleotide):
    """ Returns the complementary nucleotide

        nucleotide: a nucleotide (A, C, G, or T) represented as a string
        returns: the complementary nucleotide
    >>> get_complement('A')
    'T'
    >>> get_complement('C')
    'G'
    """
    # TODO: implement this
    if nucleotide == 'A':
        return 'T'
    elif nucleotide == 'T':
        return 'A'
    elif nucleotide == 'G':
        return 'C'
    elif nucleotide == 'C':
        return 'G'
    else:
        return


def get_reverse_complement(dna):
    """ Computes the reverse complementary sequence of DNA for the specfied DNA
        sequence

        dna: a DNA sequence represented as a string
        returns: the reverse complementary DNA sequence represented as a string
    >>> get_reverse_complement("ATGCCCGCTTT")
    'AAAGCGGGCAT'
    >>> get_reverse_complement("CCGCGTTCA")
    'TGAACGCGG'
    """
    # TODO: implement this
    revdna = ''
    for i in range(len(dna)):
        revdna = get_complement(dna[i]) + revdna
    return revdna


def rest_of_ORF(dna):
    """ Takes a DNA sequence that is assumed to begin with a start codon and returns
        the sequence up to but not including the first in frame stop codon.  If there
        is no in frame stop codon, returns the whole string.

        dna: a DNA sequence
        returns: the open reading frame represented as a string
    >>> rest_of_ORF("ATGTGAA")
    'ATG'
    >>> rest_of_ORF("ATGAGATAGG")
    'ATGAGA'
    """
    # TODO: implement this
    rest = ''
    i = 0
    while True:
        cur_codon = dna[3*i:3*i+3]
        if cur_codon == 'TAG' or cur_codon == 'TAA' or cur_codon == 'TGA':
            break
        rest += cur_codon
        i += 1
        if 3*i+3 > len(dna):
            return dna
    return rest
</code></pre>
<br>
<h4> Instead of doing 3 * i + 3 every where, you could increment i by 3 and simply do i: i + 3. </h4>
</br>
<pre><code class="language-python">

def find_all_ORFs_oneframe(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence and returns
        them as a list.  This function should only find ORFs that are in the default
        frame of the sequence (i.e. they start on indices that are multiples of 3).
        By non-nested we mean that if an ORF occurs entirely within
        another ORF, it should not be included in the returned list of ORFs.

        dna: a DNA sequence
        returns: a list of non-nested ORFs
    >>> find_all_ORFs_oneframe("ATGCATGAATGTAGATAGATGTGCCC")
    ['ATGCATGAATGTAGA', 'ATGTGCCC']
    """
    # TODO: implement this
    i = 0
    ORFs_oneframe = []
    while 3*i+3 < len(dna):
        cur_codon = dna[3*i:3*i+3]
        if cur_codon == 'ATG':
            ORFs_oneframe.append(rest_of_ORF(dna[3*i:]))
            i += len(ORFs_oneframe[-1])/3
        i += 1
    return ORFs_oneframe


def find_all_ORFs(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence in all 3
        possible frames and returns them as a list.  By non-nested we mean that if an
        ORF occurs entirely within another ORF and they are both in the same frame,
        it should not be included in the returned list of ORFs.

        dna: a DNA sequence
        returns: a list of non-nested ORFs

    >>> find_all_ORFs("ATGCATGAATGTAG")
    ['ATGCATGAATGTAG', 'ATGAATGTAG', 'ATG']
    """
    # TODO: implement this
    ORFs = []
    ORFs += find_all_ORFs_oneframe(dna)
    ORFs += find_all_ORFs_oneframe(dna[1:])
    ORFs += find_all_ORFs_oneframe(dna[2:])
    return ORFs


def find_all_ORFs_both_strands(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence on both
        strands.

        dna: a DNA sequence
        returns: a list of non-nested ORFs
    >>> find_all_ORFs_both_strands("ATGCGAATGTAGCATCAAA")
    ['ATGCGAATG', 'ATGCTACATTCGCAT']
    """
    # TODO: implement this
    ORFs_both_strands = []
    ORFs_both_strands += find_all_ORFs(dna)
    ORFs_both_strands += find_all_ORFs(get_reverse_complement(dna))
    return ORFs_both_strands


def longest_ORF(dna):
    """ Finds the longest ORF on both strands of the specified DNA and returns it
        as a string
    >>> longest_ORF("ATGCGAATGTAGCATCAAA")
    'ATGCTACATTCGCAT'
    """
    # TODO: implement this
    ORFs = find_all_ORFs_both_strands(dna)
    num_ORFs = len(ORFs)
    longest = ORFs[0]
    for i in range(num_ORFs):
        if len(ORFs[i]) > len(longest):
            longest = ORFs[i]
    return longest


def longest_ORF_noncoding(dna, num_trials):
    """ Computes the maximum length of the longest ORF over num_trials shuffles
        of the specfied DNA sequence

        dna: a DNA sequence
        num_trials: the number of random shuffles
        returns: the maximum length longest ORF """
    # TODO: implement this
    max_len = 0
    for i in range(num_trials):
        ORF = longest_ORF(shuffle_string(dna))
        if len(ORF) > max_len:
            max_len = len(ORF)
    return max_len


def coding_strand_to_AA(dna):
    """ Computes the Protein encoded by a sequence of DNA.  This function
        does not check for start and stop codons (it assumes that the input
        DNA sequence represents an protein coding region).

        dna: a DNA sequence represented as a string
        returns: a string containing the sequence of amino acids encoded by the
                 the input DNA fragment

        >>> coding_strand_to_AA("ATGCGA")
        'MR'
        >>> coding_strand_to_AA("ATGCCCGCTTT")
        'MPA'
    """
    # TODO: implement this
    aa_from_dna = ''
    for i in range(len(dna)/3):
        aa_from_dna += aa_table[dna[3*i:3*i+3]]
    return aa_from_dna

def gene_finder(dna):
    """ Returns the amino acid sequences that are likely coded by the specified dna

        dna: a DNA sequence
        returns: a list of all amino acid sequences coded by the sequence dna.
    """
    # TODO: implement this
    ORFs = find_all_ORFs_both_strands(dna)
    ORFs_with_threshold = []
    AAs = []
    for i in range(len(ORFs)):
        if len(ORFs[i]) > threshold:
            ORFs_with_threshold.append(ORFs[i])
    for j in range(len(ORFs_with_threshold)):
        AAs.append(coding_strand_to_AA(ORFs_with_threshold[j]))
    return AAs



if __name__ == "__main__":
    import doctest
    doctest.testmod()


dna = load_seq("./data/X73525.fa")
threshold = longest_ORF_noncoding(dna, 1500)
print gene_finder(dna, threshold)
</code></pre>
<br>
<h4> These should be run in __name__ == "__main__" to prevent it from running on importing this file!</h4>
</br>
<pre><code class="language-python">
</code></pre>