<h2> Gene Finder Final </h2>

<blockquote>
kailevy/gene_finder.py
</blockquote>
<h3> Great Job! I have a few comments about some parts of your code below, but none were critical enough to really affect your grade. They are things you should keep in mind for the future that will cost you points. Also, for future reference, remember to comment your code where necessary. </h3>
<pre><code class="language-python">
# -*- coding: utf-8 -*-
"""
Created on Sun Feb  2 11:24:42 2014

@author: Kai Levy

"""

# you may find it useful to import these variables (although you are not required to use them)
from amino_acids import aa, codons, aa_table
import random
from load import load_seq

def shuffle_string(s):
    """ Shuffles the characters in the input string
        NOTE: this is a helper function, you do not have to modify this in any way """
    return ''.join(random.sample(s,len(s)))

### YOU WILL START YOUR IMPLEMENTATION FROM HERE DOWN ##

def get_complement(nucleotide):
    """ Returns the complementary nucleotide

        nucleotide: a nucleotide (A, C, G, or T) represented as a string
        returns: the complementary nucleotide
    >>> get_complement('A')
    'T'
    >>> get_complement('C')
    'G'
    >>> get_complement('T')
    'A'
    >>> get_complement('G')
    'C'
    """
    opposite = {'A':'T','T':'A','C':'G','G':'C'}
    return opposite[nucleotide]

def get_reverse_complement(dna):
    """ Computes the reverse complementary sequence of DNA for the specfied DNA
        sequence

        dna: a DNA sequence represented as a string
        returns: the reverse complementary DNA sequence represented as a string
    >>> get_reverse_complement("ATGCCCGCTTT")
    'AAAGCGGGCAT'
    >>> get_reverse_complement("CCGCGTTCA")
    'TGAACGCGG'
    """
    dna_rev = dna[::-1]
    rev = [get_complement(c) for c in dna_rev]
    return ''.join(rev)
</code></pre>
<br>
<h4> Doing it this way results in 2 loops throughout nucleotide because you use a list comprehension (1 loop) to get the complements and then you use join (which goes through rev again).</h4>
</br>
<pre><code class="language-python">
def rest_of_ORF(dna):
    """ Takes a DNA sequence that is assumed to begin with a start codon and returns
        the sequence up to but not including the first in frame stop codon.  If there
        is no in frame stop codon, returns the whole string.

        dna: a DNA sequence
        returns: the open reading frame represented as a string
    >>> rest_of_ORF("ATGTGAA")
    'ATG'
    >>> rest_of_ORF("ATGAGATAGG")
    'ATGAGA'
    """
    stoppers = ['TAG','TAA','TGA']
    rest = ''
    for i in range(0,len(dna),3):
        if (dna[i:i+3] in stoppers):
            return rest
        rest += dna[i:i+3]
    return rest

def find_all_ORFs_oneframe(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence and returns
        them as a list.  This function should only find ORFs that are in the default
        frame of the sequence (i.e. they start on indices that are multiples of 3).
        By non-nested we mean that if an ORF occurs entirely within
        another ORF, it should not be included in the returned list of ORFs.

        dna: a DNA sequence
        returns: a list of non-nested ORFs
    >>> find_all_ORFs_oneframe("ATGCATGAATGTAGATAGATGTGCCC")
    ['ATGCATGAATGTAGA', 'ATGTGCCC']
    >>> find_all_ORFs_oneframe("ATGCGAATGTAGCATCAAA")
    ['ATGCGAATG']
    """
    start = 'ATG'
    i = 0
    orfs = []
    while i < len(dna):
        if dna[i:i+3] == start:
            orf_temp = rest_of_ORF(dna[i:])
            orfs.append(orf_temp)
            i += len(orf_temp) - 3
        i += 3
    return orfs
</code></pre>
<br>
<h4> You can use some comments here explaining what you're doing. (why do you - 3, what does i mean) </h4>
</br>
<pre><code class="language-python">
def find_all_ORFs(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence in all 3
        possible frames and returns them as a list.  By non-nested we mean that if an
        ORF occurs entirely within another ORF and they are both in the same frame,
        it should not be included in the returned list of ORFs.

        dna: a DNA sequence
        returns: a list of non-nested ORFs

    >>> find_all_ORFs("ATGCATGAATGTAG")
    ['ATGCATGAATGTAG', 'ATGAATGTAG', 'ATG']
    >>> find_all_ORFs("AATGCATGAATGTAG")
    ['ATG', 'ATGCATGAATGTAG', 'ATGAATGTAG']
    """
    return find_all_ORFs_oneframe(dna) + find_all_ORFs_oneframe(dna[1:]) + find_all_ORFs_oneframe(dna[2:])

def find_all_ORFs_both_strands(dna):
    """ Finds all non-nested open reading frames in the given DNA sequence on both
        strands.

        dna: a DNA sequence
        returns: a list of non-nested ORFs
    >>> find_all_ORFs_both_strands("ATGCGAATGTAGCATCAAA")
    ['ATGCGAATG', 'ATGCTACATTCGCAT']
    """
    return find_all_ORFs(dna) + find_all_ORFs(get_reverse_complement(dna))

def longest_ORF(dna):
    """ Finds the longest ORF on both strands of the specified DNA and returns it
        as a string
    >>> longest_ORF("ATGCGAATGTAGCATCAAA")
    'ATGCTACATTCGCAT'
    """
    all_ORFs = find_all_ORFs_both_strands(dna)
    if len(all_ORFs):
        return max(all_ORFs, key=len)

def longest_ORF_noncoding(dna, num_trials):
    """ Computes the maximum length of the longest ORF over num_trials shuffles
        of the specfied DNA sequence

        dna: a DNA sequence
        num_trials: the number of random shuffles
        returns: the maximum length longest ORF """
    longestOrfs = []
    for i in range(num_trials):
        longestOrfs.append(longest_ORF(shuffle_string(dna)))
    return len(max(longestOrfs, key = len))

def coding_strand_to_AA(dna):
    """ Computes the Protein encoded by a sequence of DNA.  This function
        does not check for start and stop codons (it assumes that the input
        DNA sequence represents an protein coding region).

        dna: a DNA sequence represented as a string
        returns: a string containing the sequence of amino acids encoded by the
                 the input DNA fragment

        >>> coding_strand_to_AA("ATGCGA")
        'MR'
        >>> coding_strand_to_AA("ATGCCCGCTTT")
        'MPA'
        >>> coding_strand_to_AA("ATGCGAATGTAGCATCAAA")
        'MRM|HQ'
    """
    aa = [aa_table[dna[i:i+3]] for i in range(0,len(dna)-2,3)]
    return ''.join(aa)
</code></pre>
<br>
<h4> Doing it this way results in 2 loops throughout nucleotide because you use a list comprehension (1 loop) to get the complements and then you use join (which goes through aa again).</h4>
</br>
<pre><code class="language-python">

def gene_finder(dna):
    """ Returns the amino acid sequences that are likely coded by the specified dna

        dna: a DNA sequence
        returns: a list of all amino acid sequences coded by the sequence dna.
    """
    thresh = longest_ORF_noncoding(dna,1500)
    orfs = [s for s in find_all_ORFs_both_strands(dna) if len(s) > thresh]
    return [coding_strand_to_AA(s) for s in orfs]


if __name__ == "__main__":
    fname = 'salmonella2.txt'
    dna =  load_seq("./data/X73525.fa")
    with open(fname,'w') as fout:
        for s in gene_finder(dna):
            fout.write(s + '\n')
    # import doctest
    # doctest.testmod()
</code></pre>
